### PR TITLE
Fix concurrent actor starting too many threads.

### LINKED ISSUE #14894: concurrent actor starts too many threads
### What is the problem?

An actor of max_concurrency = N should only start at most N threads. However, due to [this bug](https://github.com/ray-project/ray/pull/7335#pullrequestreview-617272911), the actual behavior is starting N threads per caller.

### Reproduction (REQUIRED)

```
def test_actor_max_concurrency(ray_start_regular):
    """
    Test that an actor of max_concurrency=N should only run
    N tasks at most concurrently.
    """
    CONCURRENCY = 5

    @ray.remote
    class ConcurentActor:
        def __init__(self):
            self.concurrent_tasks = 0
            self.lock = threading.Lock()
            self.max_concurrency = 0

        def call(self):
            with self.lock:
                self.concurrent_tasks += 1
                self.max_concurrency = max(self.max_concurrency,
                                           self.concurrent_tasks)
            time.sleep(0.5)
            with self.lock:
                self.concurrent_tasks -= 1

        def get_max_concurrency(self):
            return self.max_concurrency

    @ray.remote
    def call(actor):
        ray.get([actor.call.remote() for _ in range(CONCURRENCY * 2)])
        return

    actor = ConcurentActor.options(max_concurrency=CONCURRENCY).remote()
    # Start N tasks to call this actor concurrently.
    ray.get([call.remote(actor) for _ in range(CONCURRENCY)])

    assert ray.get(actor.get_max_concurrency.remote()) <= CONCURRENCY
```

If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

### PR DESCRIPTION
**Pull request by raulchen**
## Why are these changes needed?

Fixes #14894

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/core_worker/transport/direct_actor_transport.cc [hunk 1]
```diff
@@ -563,4 +567,23 @@ bool CoreWorkerDirectTaskReceiver::CancelQueuedNormalTask(TaskID task_id) {
...
+  if (max_concurrency_ > 1) {
+    max_concurrency_ = max_concurrency;
+    is_asyncio_ = is_asyncio;
+    if (is_asyncio_) {
+      RAY_LOG(INFO) << "Creating new thread pool of size " << max_concurrency;
+      fiber_state_.reset(new FiberState(max_concurrency));
+    } else {
+      RAY_LOG(INFO) << "Setting actor as async with max_concurrency=" << max_concurrency
+                    << ", creating new fiber thread.";
+      pool_.reset(new BoundedExecutor(max_concurrency));
```
Comment (edoakes) [inline src/ray/core_worker/transport/direct_actor_transport.cc:584]: should we `assert` here that the pool is only ever created once?

File: python/ray/tests/test_basic_2.py [hunk 2]
```diff
@@ -558,6 +558,44 @@ def add(self, x):
     assert r1 == r2 == r3
 
 
+def test_actor_max_concurrency(ray_start_regular):
```
Comment (edoakes) [inline python/ray/tests/test_basic_2.py:561]: Can we also test for the number of threads actually started? You could make a bunch of calls that return a thread ID (using https://docs.python.org/3/library/threading.html#threading.current_thread) and assert the set is the expected length. We should also do it for multiple callers (to catch the bug you identified).

File: python/ray/tests/test_basic_2.py [hunk 3]
```diff
@@ -558,6 +558,44 @@ def add(self, x):
...
+            self.concurrent_tasks = 0
+            self.lock = threading.Lock()
+            self.max_concurrency = 0
+
+        def call(self):
+            with self.lock:
+                self.concurrent_tasks += 1
+                self.max_concurrency = max(self.max_concurrency,
+                                           self.concurrent_tasks)
+            time.sleep(0.5)
```
Comment (edoakes) [inline python/ray/tests/test_basic_2.py:580]: let's avoid sleeping here please. you can use the `ray.test_utils.SignalActor` to control when to block/unblock explicitly

**Review by edoakes — COMMENTED** [review]
LGTM but see comments on tests

**Comment by raulchen** [comment]
@edoakes thanks, all comments are addressed

**Review by edoakes — APPROVED** [review]

**Review by edoakes — APPROVED** [review]
@raulchen looks good, but tests are failing:
https://travis-ci.com/github/ray-project/ray/jobs/494776823#L1076

**Pull request merged**